fbgemm_gpu/src/jagged_tensor_ops.cu

```diff
@@ -143,6 +143,381 @@ __launch_bounds__(kMaxThreads) void jagged_dense_elementwise_dense_output_kernel
   }
 }
 
+
+// brute-force offset generation kernel - not used, performance can be tuned
+// can be used to overcome shared memory size constraints
+template <typename index_t>
+__global__ void dense_offset_to_jagged_map(
+  index_t* dense_offsets,
+  index_t* dense_to_jagged_map,
+  int32_t num_offsets,
+  int32_t jagged_dim_stride,
+  int32_t e_dim_stride) {
+
+  int32_t idx = threadIdx.x + blockIdx.x * blockDim.x;
+  if (idx < (num_offsets-1)) {
+    index_t start_offset = dense_offsets[idx];
+    index_t nnz = dense_offsets[idx + 1] - start_offset;
+
+    index_t stride = (idx * jagged_dim_stride);
+
+    for (index_t k = 0; k < nnz; k++) {
+      dense_to_jagged_map[start_offset + k] = stride + (k * e_dim_stride);
+    }
+  }
+}
+
+template <typename index_t>
+__global__ void dense_offset_to_jagged_map_opt_kernel_(
+  index_t* dense_offsets,
+  index_t* dense_to_jagged_map,
+  int32_t num_offsets,
+  int32_t total_jagged_rows,
+  int32_t jagged_dim_stride,
+  int32_t e_dim_stride) {
+
+  extern __shared__ index_t smem_dense_offsets[];
+
+  // should just stream from L2 to all smem's
+  for (int32_t i = threadIdx.x; i < num_offsets; i += blockDim.x) {
+    smem_dense_offsets[i] = dense_offsets[i];
+  }
+  __syncthreads();
+
+  for (index_t idx = (threadIdx.x + blockIdx.x * blockDim.x); idx < total_jagged_rows;
+          idx += (blockDim.x * gridDim.x)) {
+    // search for idx in smem_dense_offsets
+    index_t mid = num_offsets / 2;
+    index_t lower = 0, upper = (num_offsets-1); // offset shoud have atleast 2 values
+    index_t res = 0;
+
+    while (lower <= upper) {
+      if (idx >= smem_dense_offsets[mid]) {
+        lower = mid + 1;
+        res = mid;
+      } else {
+        upper = mid - 1;
+      }
+      mid = (lower + upper) / 2;
+    }
+    // warp sync ?
+    index_t ix = idx - smem_dense_offsets[res];
+    index_t stride = (res * jagged_dim_stride); // res is row in dense-0th dim
```

Comment:
Strides are manually used vs. tensor accessor. Does it make a difference in perf?

Reply:
It shouldn't make any difference in perf as long as accessor is also adding same amount of instructions. I used explicit calculations as it makes the code self explanatory.